pkg/cmd/adm/restart.go

```diff
@@ -1,157 +1,200 @@
-		}
-		return fmt.Errorf("at least one deployment name is required, include one or more of the above deployments to restart")
+	cl, err := ctx.NewClient(cfg.Token, cfg.ServerAPI)
+	if err != nil {
+		return err
 	}
-	deploymentName := deployments[0]
 
 	if !ctx.AskForConfirmation(
-		ioutils.WithMessagef("restart the deployment '%s' in namespace '%s'", deploymentName, cfg.OperatorNamespace)) {
+		ioutils.WithMessagef("restart all the deployments in the cluster  '%s' and namespace '%s' \n", clusterName, cfg.OperatorNamespace)) {
 		return nil
 	}
-	return restartDeployment(ctx, cl, cfg.OperatorNamespace, deploymentName)
+
+	return restartDeployment(ctx, cl, cfg.OperatorNamespace, factory, ioStreams)
 }
 
-func restartDeployment(ctx *clicontext.CommandContext, cl runtimeclient.Client, ns string, deploymentName string) error {
-	namespacedName := types.NamespacedName{
-		Namespace: ns,
-		Name:      deploymentName,
-	}
+func restartDeployment(ctx *clicontext.CommandContext, cl runtimeclient.Client, ns string, f cmdutil.Factory, ioStreams genericclioptions.IOStreams) error {
+	ctx.Printlnf("Fetching the current OLM and non-OLM deployments of the operator in %s namespace", ns)
 
-	originalReplicas, err := scaleToZero(cl, namespacedName)
+	olmDeploymentList, nonOlmDeploymentlist, err := getExistingDeployments(ctx, cl, ns)
 	if err != nil {
-		if apierrors.IsNotFound(err) {
-			ctx.Printlnf("\nERROR: The given deployment '%s' wasn't found.", deploymentName)
-			return printExistingDeployments(ctx, cl, ns)
-		}
-		return err
-	}
-	ctx.Println("The deployment was scaled to 0")
-	if err := scaleBack(ctx, cl, namespacedName, originalReplicas); err != nil {
-		ctx.Printlnf("Scaling the deployment '%s' in namespace '%s' back to '%d' replicas wasn't successful", originalReplicas)
-		ctx.Println("Please, try to contact administrators to scale the deployment back manually")
 		return err
 	}
 
-	ctx.Printlnf("The deployment was scaled back to '%d'", originalReplicas)
+	if len(olmDeploymentList.Items) == 0 {
+		ctx.Printlnf("No OLM based deployment restart happened as Olm deployment found in namespace %s is 0", ns)
+	} else {
+		for _, olmDeployment := range olmDeploymentList.Items {
+			ctx.Printlnf("Proceeding to delete the Pods of %v", olmDeployment)
+
+			if err := deleteAndWaitForPods(ctx, cl, olmDeployment, f, ioStreams); err != nil {
+				return err
+			}
+		}
+	}
+	if len(nonOlmDeploymentlist.Items) != 0 {
+		for _, nonOlmDeployment := range nonOlmDeploymentlist.Items {
+
+			ctx.Printlnf("Proceeding to restart the non-OLM deployment %v", nonOlmDeployment)
+
+			if err := restartNonOlmDeployments(ctx, nonOlmDeployment, f, ioStreams); err != nil {
+				return err
+			}
+			//check the rollout status
+			ctx.Printlnf("Checking the status of the rolled out deployment %v", nonOlmDeployment)
+			if err := checkRolloutStatus(ctx, f, ioStreams, "provider=codeready-toolchain"); err != nil {
+				return err
```

Comment:
Let's try to simplify this. You can group some logic together and make it more generic. As I mentioned in the other comment, when we are getting the the list of deployments, we don't have to care if they are OLM or non-OLM based - we just:
* list them based on the labels
* check if there are the operator deployments (if not then fail)
* filter out the autoscaler one

then you can iterate over each of the deployments:
1. check if the deployment has some owner-reference
    i. if it has then delete the pods
    ii. if not then use the rollout restart logic
2. wait for the rollout status

with this approach, you can have the logic simpler, with only one for-loop, only one if-else statement, and call the `checkRolloutStatus` function only from one place

Reply:
i have tried to simplify few things